url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/game/SOCDevCardSet.java

```diff
@@ -0,0 +1,232 @@
+        return devCards[age][ctype];
+    }
+    
+    /** Added by Katie
+     * @return total number of each type of card
+     */
+    public int[] getTotalTypes(){
+    	int[] cards= new int[SOCDevCardConstants.MAX_KNOWN+1];
+    	for (int i = SOCDevCardConstants.MIN; i<= SOCDevCardConstants.MAX_KNOWN; i++){
+    		cards[i]= devCards[OLD][i]+devCards[NEW][i];
+    	}
+    	return cards;
+    }
+
+    /**
+     * @return the total number of development cards
+     */
+    public int getTotal()
+    {
+        int sum = 0;
+
+        for (int i = SOCDevCardConstants.MIN;
+                i < SOCDevCardConstants.MAXPLUSONE; i++)
+        {
+            sum += (devCards[OLD][i] + devCards[NEW][i]);
+        }
+
+        return sum;
+    }
+
+    /**
+     * set the amount of a type of card
+     *
+     * @param age   either {@link #OLD} or {@link #NEW}
+     * @param ctype the type of development card, at least
+     *              {@link SOCDevCardConstants#MIN} and less than {@link SOCDevCardConstants#MAXPLUSONE}
+     * @param amt   the amount
+     */
+    public void setAmount(int amt, int age, int ctype)
+    {
+        devCards[age][ctype] = amt;
+    }
+
+    /**
+     * add an amount to a type of card
+     *
+     * @param age   either {@link #OLD} or {@link #NEW}
+     * @param ctype the type of development card, at least
+     *              {@link SOCDevCardConstants#MIN} and less than {@link SOCDevCardConstants#MAXPLUSONE}
+     * @param amt   the amount
+     */
+    public void add(int amt, int age, int ctype)
+    {
+        devCards[age][ctype] += amt;
+    }
+
+    /**
+     * subtract an amount from a type of card
+     *
+     * @param age   either {@link #OLD} or {@link #NEW}
+     * @param ctype the type of development card, at least
+     *              {@link SOCDevCardConstants#MIN} and less than {@link SOCDevCardConstants#MAXPLUSONE}
+     * @param amt   the amount
+     */
+    public void subtract(int amt, int age, int ctype)
+    {
+        if (amt <= devCards[age][ctype])
+        {
+            devCards[age][ctype] -= amt;
+        }
+        else
+        {
+            devCards[age][ctype] = 0;
+            devCards[age][SOCDevCardConstants.UNKNOWN] -= amt;
+        }
+    }
+
+    /**
+     * @return the number of victory point cards in
+     *         this set
+     */
+    public int getNumVPCards()
+    {
+        int sum = 0;
+
+        sum += devCards[OLD][SOCDevCardConstants.CAP];
+        sum += devCards[OLD][SOCDevCardConstants.LIB];
+        sum += devCards[OLD][SOCDevCardConstants.UNIV];
+        sum += devCards[OLD][SOCDevCardConstants.TEMP];
+        sum += devCards[OLD][SOCDevCardConstants.TOW];
+        sum += devCards[NEW][SOCDevCardConstants.CAP];
+        sum += devCards[NEW][SOCDevCardConstants.LIB];
+        sum += devCards[NEW][SOCDevCardConstants.UNIV];
+        sum += devCards[NEW][SOCDevCardConstants.TEMP];
+        sum += devCards[NEW][SOCDevCardConstants.TOW];
+
+        return sum;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 180:193, 204:217.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.